exonum/src/api/backends/actix.rs

```diff
@@ -131,116 +132,143 @@ impl ResponseError for api::Error {
             api::Error::Unauthorized => HttpResponse::Unauthorized().finish(),
         }
     }
 }
 
-impl<Q, I, F> From<NamedWith<Q, I, api::Result<I>, F, Immutable>> for RequestHandler
-where
-    F: Fn(Q) -> api::Result<I> + 'static + Send + Sync + Clone,
-    Q: DeserializeOwned + 'static,
-    I: Serialize + 'static,
-{
-    fn from(f: NamedWith<Q, I, api::Result<I>, F, Immutable>) -> Self {
-        let handler = f.inner.handler;
-        let index = move |request: HttpRequest| -> FutureResponse {
-            let future = Query::from_request(&request, &Default::default())
-                .map(Query::into_inner)
-                .and_then(|query| handler(query).map_err(From::from))
-                .and_then(|value| Ok(HttpResponse::Ok().json(value)))
-                .into_future();
-            Box::new(future)
+/// Creates a `HttpResponse` object from the provided JSON value.
+/// Depending on the `actuality` parameter value, the warning about endpoint
+/// being deprecated can be added.
+fn json_response<T: Serialize>(actuality: Actuality, json_value: T) -> HttpResponse {
+    let mut response = HttpResponse::Ok();
+
+    if let Actuality::Deprecated(ref discontinued_on) = actuality {
+        // There is a proposal for creating special deprecation header within HTTP,
+        // but currently it's only a draft. So the conventional way to notify API user
+        // about endpoint deprecation is setting the `Warning` header.
+        let expiration_note = match discontinued_on {
+            Some(date) => format!(
+                "The old API is maintained until {}.",
+                date.format("%Y-%m-%d")
+            ),
+            None => "Currently there is no specific date for disabling this endpoint.".into(),
         };
 
-        Self {
-            name: f.name,
-            method: actix_web::http::Method::GET,
-            inner: Arc::from(index) as Arc<RawHandler>,
+        let warning_text = format!(
+            "Deprecated API: This endpoint is deprecated, \
+             see the documentation to find an alternative. \
+             {}",
+            expiration_note
+        );
+
+        let warning_string = create_warning_header(&warning_text);
+
+        response.header(header::WARNING, warning_string);
+    }
+
+    response.json(json_value)
+}
+
+/// Formats warning string according to the following format:
+/// "<warn-code> <warn-agent> \"<warn-text>\" [<warn-date>]"
+/// <warn-code> in our case is 299, which means a miscellaneous persistent warning.
+/// <warn-agent> is optional, so we set it to "-".
+/// <warn-text> is a warning description, which is taken as an only argument.
+/// <warn-date> is not required.
+/// For details you can see RFC 7234, section 5.5: Warning.
+fn create_warning_header(warning_text: &str) -> String {
+    const WARNING_NUMBER: u16 = 299;
+    const WARNING_AGENT: &str = "-";
+
+    format!("{} {} \"{}\"", WARNING_NUMBER, WARNING_AGENT, warning_text)
```

Comment:
Micro-nit: I wonder whether constants are inlined by the compiler here. Could give a (microscopic) performance boost.